controllers/registry_controller.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	config "github.com/openshift/api/config/v1"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/cluster"
+	"github.com/openshift/windows-machine-config-operator/pkg/condition"
+	"github.com/openshift/windows-machine-config-operator/pkg/registries"
+)
+
+const (
+	// RegistryController is the name of this controller in logs and other outputs.
+	RegistryController = "registry"
+)
+
+// registryReconciler holds the info required to reconcile image registry settings on Windows nodes
+type registryReconciler struct {
+	instanceReconciler
+}
+
+// NewRegistryReconciler returns a pointer to a new registryReconciler
+func NewRegistryReconciler(mgr manager.Manager, clusterConfig cluster.Config,
+	watchNamespace string) (*registryReconciler, error) {
+	clientset, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, fmt.Errorf("error creating kubernetes clientset: %w", err)
+	}
+
+	return &registryReconciler{
+		instanceReconciler: instanceReconciler{
+			client:             mgr.GetClient(),
+			log:                ctrl.Log.WithName("controllers").WithName(RegistryController),
+			k8sclientset:       clientset,
+			clusterServiceCIDR: clusterConfig.Network().GetServiceCIDR(),
+			watchNamespace:     watchNamespace,
+			recorder:           mgr.GetEventRecorderFor(RegistryController),
+		},
+	}, nil
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which reads that state of the cluster for objects
+// related to image registry config and aims to move the current state of the cluster closer to the desired state.
+func (r *registryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	r.log = r.log.WithValues(RegistryController, req.NamespacedName)
+	// Prevent WMCO upgrades while registry configuration settings are being processed
+	if err := condition.MarkAsBusy(r.client, r.watchNamespace, r.recorder, RegistryController); err != nil {
+		return ctrl.Result{}, err
+	}
+	defer func() {
+		err = markAsFreeOnSuccess(r.client, r.watchNamespace, r.recorder, RegistryController, result.Requeue, err)
+	}()
```

Comment:
Do upgrades really have to be prevented by this controller?
The whole containerd directory is being wiped on upgrade and re-created.
We should be cautious when adding this lock, as there is the small chance of deadlock when upgrading the cluster due to changes in the payload.

Reply:
I was thinking of a case where our wm or cm controllers were upgrading a node (i.e. deleting directories) and, at the same time, the registry controller is trying to write new config to the containerd dir. Wouldn't that clash?

Reply:
> Wouldn't that clash?
Potentially, but as far as I can tell the impact would be the registry controller fails and retries once the node is upgraded.

Reply:
Fair enough, I've removed this